In [1]:
import pandas as pd
import numpy as np
import os

In [2]:
path = r'Users/bridget/Downloads/Instacart Basket Analysis 11-2022'

In [6]:
# Import the 'orders_products_merged_2.pkl' dataset
df_ords_merged = pd.read_pickle(r'/Users/bridget/Downloads/Instacart Basket Analysis 11-2022/orders_products_merged_2.pkl')

Aggregating data with the agg() function 
Perform aggreations on the dataframe, grouped by the 'department_id' column

In [7]:
# Create a filter of df_ords_merged
df = df_ords_merged[:1000000]

print('Sample output of filter of df_ords_merged:')
df.sample(5)

Sample output of filter of df_ords_merged:


order_id  user_id eval_set  order_number  order_dow  \
630075   1891389   166027    prior            17          3   
454681   2368962    12346    prior            36          6   
378855   1408714   171627    prior             9          0   
916765   1863008    76122    prior            17          3   
345880   2039845   153456    prior             7          6   

        order_hour_of_day  days_since_prior_order  product_id  \
630075                 10                     2.0       38928   
454681                 10                     6.0       30450   
378855                 12                     5.0       13176   
916765                 11                     7.0       27344   
345880                 10                     4.0       13176   

        add_to_cart_order  reordered  Unnamed: 0              product_name  \
630075                  1          1       38932  0% Greek Strained Yogurt   
454681                  4          1       30452      Creamy Almond Butter   
378855                  1          1       13176    Bag of Organic Bananas   
916765                  2          0       27346      Uncured Genoa Salami   
345880                  4          1       13176    Bag of Organic Bananas   

        aisle_id  department_id  prices _merge        price_range  \
630075       120             16    12.6   both  Mid-range product   
454681        88             13     3.2   both  Low-range product   
378855        24              4    10.3   both  Mid-range product   
916765        96             20    14.8   both  Mid-range product   
345880        24              4    10.3   both  Mid-range product   

       busiest_period_of_day  
630075           Most orders  
454681           Most orders  
378855           Most orders  
916765           Most orders  
345880           Most orders

In [8]:
df_ords = pd.read_csv(r'/Users/bridget/Downloads/Instacart Basket Analysis 11-2022/02 Data/Original Data/orders.csv', index_col = False)

In [10]:
df_ords_prods_merged = pd.read_pickle(os.path.join(path, '02 Data','Prepared Data', '/Users/bridget/Downloads/Instacart Basket Analysis 11-2022/02 Data/Prepared Data/orders_products_merged.pkl'))

In [11]:
df_ords_prods_merged.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

In [12]:
df_ords.drop(columns = ['eval_set'])

order_id  user_id  order_number  order_dow  order_hour_of_day  \
0         2539329        1             1          2                  8   
1         2398795        1             2          3                  7   
2          473747        1             3          3                 12   
3         2254736        1             4          4                  7   
4          431534        1             5          4                 15   
...           ...      ...           ...        ...                ...   
3421078   2266710   206209            10          5                 18   
3421079   1854736   206209            11          4                 10   
3421080    626363   206209            12          1                 12   
3421081   2977660   206209            13          1                 12   
3421082    272231   206209            14          6                 14   

         days_since_prior_order  
0                           NaN  
1                          15.0  
2                          21.0  
3                          29.0  
4                          28.0  
...                         ...  
3421078                    29.0  
3421079                    30.0  
3421080                    18.0  
3421081                     7.0  
3421082                    30.0  

[3421083 rows x 6 columns]

In [13]:
# Group the df dataframe by the 'department_id' column, then calculate the mean, min and max of the 'order_number' column
print('The mean, min and max of order_number grouped by the department_id column in df:')
df.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

The mean, min and max of order_number grouped by the department_id column in df:


order_number        
                      mean min max
department_id                     
4                18.825780   1  99
7                17.472355   1  99
13               17.993423   1  99
14               19.246334   1  99
16               19.463012   1  99
17               11.294069   1  98
19               19.305237   1  99
20               17.599636   1  99

In [14]:
# Check output of df_ords_merged
print('Sample output of of df_ords_merged:/n', df_ords_merged.shape)
df_ords_merged.sample(5)

Sample output of of df_ords_merged:
 (32404859, 18)


order_id  user_id eval_set  order_number  order_dow  \
25733056   1688876   149693    prior             1          6   
15262385    111127   115700    prior             2          0   
2455273      92853    98411    prior            84          1   
11583212   1371780    37128    prior             6          4   
4876783    3335859   154842    prior            78          3   

          order_hour_of_day  days_since_prior_order  product_id  \
25733056                 18                     NaN       39194   
15262385                 11                     9.0        8204   
2455273                  22                     3.0        4957   
11583212                 16                    23.0       13984   
4876783                   7                     2.0       45007   

          add_to_cart_order  reordered  Unnamed: 0  \
25733056                  4          0       39198   
15262385                 13          0        8204   
2455273                   6          1        4957   
11583212                  7          1       13984   
4876783                   4          1       45011   

                                               product_name  aisle_id  \
25733056                         Bean & Rice Chips sea salt       107   
15262385   Traditional Favorites Tomato & Basil Pasta Sauce         9   
2455273   Total 2% Lowfat Greek Strained Yogurt With Blu...       120   
11583212                                       Organic Mint        16   
4876783                                    Organic Zucchini        83   

          department_id  prices _merge        price_range  \
25733056             19     3.8   both  Low-range product   
15262385              9    10.5   both  Mid-range product   
2455273              16     9.8   both  Mid-range product   
11583212              4     5.4   both  Mid-range product   
4876783               4     7.9   both  Mid-range product   

         busiest_period_of_day  
25733056        Average orders  
15262385           Most orders  
2455273         Average orders  
11583212           Most orders  
4876783         Average orders

In [15]:
# Group the df_ords_merged dataframe by the 'department_id' column, then calculate the mean, min and max of the 'order_number' column
print('The mean, min and max of order_number grouped by the department_id column in df_ords_merged:')
df_ords_merged.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

The mean, min and max of order_number grouped by the department_id column in df_ords_merged:


order_number        
                      mean min max
department_id                     
1                15.457838   1  99
2                17.277920   1  99
3                17.170395   1  99
4                17.811403   1  99
5                15.215751   1  99
6                16.439806   1  99
7                17.225802   1  99
8                15.340650   1  99
9                15.895474   1  99
10               20.197148   1  99
11               16.170638   1  99
12               15.887671   1  99
13               16.583536   1  99
14               16.773669   1  99
15               16.165037   1  99
16               17.665606   1  99
17               15.694469   1  99
18               19.310397   1  99
19               17.177343   1  99
20               16.473447   1  99
21               22.902379   1  99

Aggregating data with transform() function

In [16]:
# check output of df_ords_merged
print('Sample output of of df_ords_merged:/n', df_ords_merged.shape)
df_ords_merged.sample(5)

Sample output of of df_ords_merged:/n (32404859, 18)


order_id  user_id eval_set  order_number  order_dow  \
7307632    3414839    55063    prior            15          1   
2762547    1297813   106485    prior             2          5   
24454236    564512   121233    prior            12          2   
11660540    823179    45098    prior             7          2   
9288403    1280285   156982    prior             2          5   

          order_hour_of_day  days_since_prior_order  product_id  \
7307632                  20                     8.0       32299   
2762547                  15                     7.0       16521   
24454236                 10                     0.0       23763   
11660540                 11                     9.0       23576   
9288403                  13                    21.0        7877   

          add_to_cart_order  reordered  Unnamed: 0  \
7307632                   9          0       32301   
2762547                   7          0       16521   
24454236                  1          0       23764   
11660540                  7          0       23577   
9288403                   2          0        7877   

                                               product_name  aisle_id  \
7307632                                       Dijon Mustard        72   
2762547                              Walnut Halves & Pieces       117   
24454236    French Vanilla Sugar Free Liquid Coffee Creamer        53   
11660540  Sweet Greens and Lemon Vegetable and Fruit Jui...        31   
9288403                                      Chicken Thighs        49   

          department_id  prices _merge         price_range  \
7307632              13     5.9   both   Mid-range product   
2762547              19     4.7   both   Low-range product   
24454236             16     2.3   both   Low-range product   
11660540              7    14.0   both   Mid-range product   
9288403              12    23.0   both  High-range product   

         busiest_period_of_day  
7307632         Average orders  
2762547            Most orders  
24454236           Most orders  
11660540           Most orders  
9288403            Most orders

In [17]:
# Group the df_ords_merged dataframe by the 'department_id' column, then calculate the mean, min and max of the 'order_number' column
print('The mean, min and max of order_number grouped by the department_id column in df_ords_merged:')
df_ords_merged.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

The mean, min and max of order_number grouped by the department_id column in df_ords_merged:


order_number        
                      mean min max
department_id                     
1                15.457838   1  99
2                17.277920   1  99
3                17.170395   1  99
4                17.811403   1  99
5                15.215751   1  99
6                16.439806   1  99
7                17.225802   1  99
8                15.340650   1  99
9                15.895474   1  99
10               20.197148   1  99
11               16.170638   1  99
12               15.887671   1  99
13               16.583536   1  99
14               16.773669   1  99
15               16.165037   1  99
16               17.665606   1  99
17               15.694469   1  99
18               19.310397   1  99
19               17.177343   1  99
20               16.473447   1  99
21               22.902379   1  99

In [18]:
# Create new column 'max order' for placing results of max orders for each user

In [19]:
# Create a new column 'max_order' containing the maximum order number for each user_id, calculated using the groupby() and transform() functions
df_ords_merged['max_order'] = df_ords_merged.groupby(['user_id'])['order_number'].transform(np.max)

In [20]:
# Check the output of 'user_id' and 'max_order'
print('Sample user_id and max_order output:')
df_ords_merged[['user_id', 'max_order']].sample(10)

Sample user_id and max_order output:


user_id  max_order
24223144    16152         22
11861997    71734         18
13514785    26159         41
5679038    180193         82
3673118    203271         66
27551559   175939         37
26912949    14846         53
6761507     90346          5
3418605      5245         14
31146821   132244         30

Deriving columns with loc()

In [21]:
# Create the 'loyalty_flag' column based on each user's maximum order value
df_ords_merged.loc[df_ords_merged['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [22]:
df_ords_merged.loc[(df_ords_merged['max_order'] <= 40) & (df_ords_merged['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [23]:
df_ords_merged.loc[df_ords_merged['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [24]:
# Count the values in the 'loyalty_flag' column
print('The value counts in the loyalty_flag column:')
df_ords_merged['loyalty_flag'].value_counts(dropna = False)

The value counts in the loyalty_flag column:


Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

In [25]:
# Check the accuracy of the 'loyalty_flag' column with output from columns 'user_id', 'max_order', and 'loyalty_flag' column
print('Sample output from columns user_id, max_order and loyalty_flag columns:')
df_ords_merged[['user_id', 'max_order', 'loyalty_flag']].sample(30)

Sample output from columns user_id, max_order and loyalty_flag columns:


user_id  max_order      loyalty_flag
27922715   111799         78    Loyal customer
11663597    49276         46    Loyal customer
4877735    157057         29  Regular customer
1949500    183742          8      New customer
28922335    81091         20  Regular customer
23970706    76623          5      New customer
22102633   190569         13  Regular customer
6195957     82563         99    Loyal customer
29169164   115607         32  Regular customer
14842328    18544         63    Loyal customer
27666755   111004         13  Regular customer
10279273    32660         56    Loyal customer
30986955   173001         15  Regular customer
25745696   123701         38  Regular customer
17935652   131556          3      New customer
8394146    147745         44    Loyal customer
6472616    104601         58    Loyal customer
25275106   173768         35  Regular customer
16524474   177270          5      New customer
3430076     14600         19  Regular customer
9340121    127104         36  Regular customer
9063844     15995         59    Loyal customer
10335168   176742         24  Regular customer
17539150    60219         76    Loyal customer
25895190     6296          6      New customer
31093015    64405         29  Regular customer
17829935   156553         43    Loyal customer
27674259    91411          3      New customer
25236524    45236         77    Loyal customer
392883     178993         99    Loyal customer

Determine spending habits of each loyalty category in the 'loyalty_flag' column

In [26]:
# Basic statistics of df_ords_merged by the 'loyalty_flag' column, then calculate the mean, min and max of the 'prices' column
print('The mean, min and max of prices grouped by the loyalty_flag column in df_ords_merged:')
df_ords_merged.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

The mean, min and max of prices grouped by the loyalty_flag column in df_ords_merged:


prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

On average: new customers = higher prices
regular customers and loyal customers = cheapeast prices
max price = 99,999 which skews distribution 


In [27]:
# Run descriptive statistics on the 'prices' column
print('Descriptive statistics of the prices column:')
df_ords_merged['prices'].describe()

Descriptive statistics of the prices column:


count    3.240486e+07
mean     1.198023e+01
std      4.956554e+02
min      1.000000e+00
25%      4.200000e+00
50%      7.400000e+00
75%      1.130000e+01
max      9.999900e+04
Name: prices, dtype: float64

In [28]:
# Check for abnormal values in the 'prices' column
print('Check for abnormal values in the prices column:')
df_ords_merged['prices'].drop_duplicates().sort_values(ascending = False)

Check for abnormal values in the prices column:


29165516    99999.0
10030345    14900.0
18504754       25.0
25758883       24.9
28425432       24.8
             ...   
3384024         1.4
632766          1.3
5488887         1.2
5147325         1.1
60725           1.0
Name: prices, Length: 242, dtype: float64

In [29]:
# Get output of df_ords_merged when the value in the 'prices' column is 99999 or 14900
max_price = df_ords_merged.loc[df_ords_merged['prices'].isin([99999, 14900])]

In [30]:
print('Sample output of df_ords_merged when the value in the prices column is either 99999 or 14900:')
max_price.sample(5)

Sample output of df_ords_merged when the value in the prices column is either 99999 or 14900:


order_id  user_id eval_set  order_number  order_dow  \
10030760   2963082    20100    prior             4          0   
10030618   2820274    11550    prior             8          5   
10032192    724872    91578    prior            30          3   
29165857    621713   114060    prior            38          2   
10031843   1688869    73148    prior             2          4   

          order_hour_of_day  days_since_prior_order  product_id  \
10030760                 11                    14.0       21553   
10030618                 15                    15.0       21553   
10032192                 14                     6.0       21553   
29165857                 15                     3.0       33664   
10031843                 11                     5.0       21553   

          add_to_cart_order  reordered  Unnamed: 0  \
10030760                 12          0       21554   
10030618                  2          1       21554   
10032192                  5          1       21554   
29165857                 13          1       33666   
10031843                  5          1       21554   

                              product_name  aisle_id  department_id   prices  \
10030760  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0   
10030618  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0   
10032192  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0   
29165857             2 % Reduced Fat  Milk        84             16  99999.0   
10031843  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0   

         _merge         price_range busiest_period_of_day  max_order  \
10030760   both  High-range product           Most orders         19   
10030618   both  High-range product           Most orders         16   
10032192   both  High-range product           Most orders         61   
29165857   both  High-range product           Most orders         78   
10031843   both  High-range product           Most orders         10   

              loyalty_flag  
10030760  Regular customer  
10030618  Regular customer  
10032192    Loyal customer  
29165857    Loyal customer  
10031843      New customer

In [31]:
# Descriptive statistics on max_price
print('Descriptive statistics on max_price:')
max_price.describe()

Descriptive statistics on max_price:


order_id        user_id  order_number    order_dow  \
count  5.127000e+03    5127.000000   5127.000000  5127.000000   
mean   1.708876e+06  106117.185684     15.302126     2.750536   
std    9.838956e+05   58594.982659     16.691849     2.010892   
min    4.300000e+01      17.000000      1.000000     0.000000   
25%    8.642320e+05   58136.000000      4.000000     1.000000   
50%    1.693799e+06  106021.000000      9.000000     3.000000   
75%    2.567244e+06  156389.000000     20.000000     5.000000   
max    3.421047e+06  206049.000000     99.000000     6.000000   

       order_hour_of_day  days_since_prior_order    product_id  \
count        5127.000000             4776.000000   5127.000000   
mean           13.473376               11.899079  23201.815682   
std             4.137326                9.068019   4153.742747   
min             0.000000                0.000000  21553.000000   
25%            10.000000                5.000000  21553.000000   
50%            13.000000                8.000000  21553.000000   
75%            16.000000               16.000000  21553.000000   
max            23.000000               30.000000  33664.000000   

       add_to_cart_order    reordered    Unnamed: 0     aisle_id  \
count        5127.000000  5127.000000   5127.000000  5127.000000   
mean            7.986347     0.613614  23202.951824   104.732592   
std             6.868889     0.486968   4154.085720     8.231346   
min             1.000000     0.000000  21554.000000    84.000000   
25%             3.000000     0.000000  21554.000000   108.000000   
50%             6.000000     1.000000  21554.000000   108.000000   
75%            11.000000     1.000000  21554.000000   108.000000   
max            72.000000     1.000000  33666.000000   108.000000   

       department_id        prices    max_order  
count         5127.0   5127.000000  5127.000000  
mean            16.0  26485.547494    29.791691  
std              0.0  29186.636448    24.929207  
min             16.0  14900.000000     3.000000  
25%             16.0  14900.000000    12.000000  
50%             16.0  14900.000000    21.000000  
75%             16.0  14900.000000    41.000000  
max             16.0  99999.000000    99.000000

In [32]:
# Create a filter 'dairy' for records with the department_id value of 16, and then 'dairy_without_weird' to exclude records without the unusually high prices
dairy = df_ords_merged.loc[df_ords_merged['department_id'].isin([16])]
dairy_without_weird = dairy.loc[~dairy['prices'].isin([99999, 14900])]
print('Filter for records with the department_id value of 16, while excluding records without the unusually high prices:')
dairy_without_weird.sample(5)

Filter for records with the department_id value of 16, while excluding records without the unusually high prices:


order_id  user_id eval_set  order_number  order_dow  \
28230489   1515386    51548    prior             7          3   
4932042    2345908    28142    prior            11          2   
6442655    1357146   118832    prior            21          2   
592730       97406   150047    prior            35          6   
2400609    2032519    79667    prior            24          2   

          order_hour_of_day  days_since_prior_order  product_id  \
28230489                  9                    30.0       45930   
4932042                   9                    21.0       25659   
6442655                   8                    13.0       43295   
592730                   15                     6.0       49235   
2400609                  13                     8.0       17872   

          add_to_cart_order  reordered  Unnamed: 0  \
28230489                  7          0       45934   
4932042                  26          0       25660   
6442655                  19          0       43299   
592730                   15          1       49239   
2400609                   7          1       17872   

                                product_name  aisle_id  department_id  prices  \
28230489   Small Curd Low Fat Cottage Cheese       108             16     8.4   
4932042                 Organic Coconut Milk        91             16     4.1   
6442655      Unsweetened Vanilla Almond Milk        91             16    11.2   
592730                   Organic Half & Half        53             16     1.8   
2400609   Total 2% Lowfat Plain Greek Yogurt       120             16    10.4   

         _merge        price_range busiest_period_of_day  max_order  \
28230489   both  Mid-range product           Most orders         45   
4932042    both  Low-range product           Most orders         18   
6442655    both  Mid-range product        Average orders         26   
592730     both  Low-range product           Most orders         50   
2400609    both  Mid-range product           Most orders         42   

              loyalty_flag  
28230489    Loyal customer  
4932042   Regular customer  
6442655   Regular customer  
592730      Loyal customer  
2400609     Loyal customer

In [33]:
# Check if filter is working properly
print('Maximum price in the dairy_without_weird filter:')
dairy_without_weird['prices'].max()

Maximum price in the dairy_without_weird filter:


15.0

In [34]:
# Create a filter for products that contain 'Reduced Fat', '2' and 'Milk' in the name, in dairy_without_weird
reduced_fat_milk = dairy_without_weird.query('product_name.str.contains("Reduced Fat") & product_name.str.contains("2") & product_name.str.contains("Milk")')

In [35]:
print('Sample output from reduced_fat_milk:')
reduced_fat_milk.sample(30)

Sample output from reduced_fat_milk:


order_id  user_id eval_set  order_number  order_dow  \
16009529    164142    60270    prior            16          3   
15999021    543585    14351    prior            19          1   
16031053    803127   152358    prior             8          2   
14557408    180456   134431    prior            13          0   
16345601   1796330   140007    prior            21          1   
20819459   2326235     9143    prior            32          2   
11699654   3174052    80189    prior            16          4   
12861493   2907208    72153    prior            19          5   
12873696    721689   184064    prior            22          6   
16007504   1314394    50582    prior            30          1   
12861147   1823749    68968    prior            15          2   
14553652   2266973   110388    prior             3          0   
16037119    992313   176037    prior             9          6   
11697707    863041    56689    prior            14          1   
16019215   3152779   101343    prior            78          2   
16344421    352706   122847    prior            18          1   
20821088   2346691    57983    prior            32          1   
15997135    767286     6577    prior            32          5   
14538187    703327    23734    prior            34          2   
16338638    818353    47562    prior            47          1   
14564524   3243643   170467    prior             4          5   
15475159   2088294    13662    prior             6          1   
11697912   1642250    58575    prior            15          0   
14539766    143075    33675    prior             8          6   
14536151    316875    13225    prior            15          2   
14553320   2631758   108790    prior            22          1   
12856450   1664934    29221    prior            11          5   
8317131    1263690   179827    prior            53          5   
14549879   1927714    90336    prior             1          6   
15475674    914727    48079    prior             1          6   

          order_hour_of_day  days_since_prior_order  product_id  \
16009529                 22                    19.0        5785   
15999021                 13                     3.0        5785   
16031053                 11                     4.0        5785   
14557408                 18                    27.0       23909   
16345601                  7                     8.0       40174   
20819459                 16                     7.0        2962   
11699654                  9                     2.0       32478   
12861493                 17                     7.0        1940   
12873696                 18                     1.0        1940   
16007504                 16                     7.0        5785   
12861147                 11                     5.0        1940   
14553652                 15                    29.0       23909   
16037119                 10                     6.0        5785   
11697707                 12                    10.0       32478   
16019215                 15                     4.0        5785   
16344421                 11                     3.0       40174   
20821088                 10                     7.0        2962   
15997135                 14                     7.0        5785   
14538187                 15                     8.0       23909   
16338638                  7                     5.0       40174   
14564524                  7                    10.0       23909   
15475159                  8                     4.0       22610   
11697912                  0                    15.0       32478   
14539766                 21                    16.0       23909   
14536151                 12                    30.0       23909   
14553320                 13                     6.0       23909   
12856450                  9                     4.0        1940   
8317131                   9                     3.0       33572   
14549879                  9                     NaN

In [36]:
# Quick statistics in reduced_fat_milk
print('Quick statistics in reduced_fat_milk:')
reduced_fat_milk.describe()

Quick statistics in reduced_fat_milk:


order_id        user_id   order_number      order_dow  \
count  2.022380e+05  202238.000000  202238.000000  202238.000000   
mean   1.711414e+06  103040.384151      18.528684       2.748702   
std    9.876774e+05   59620.791055      18.288088       2.022770   
min    5.000000e+00      11.000000       1.000000       0.000000   
25%    8.586890e+05   51496.000000       5.000000       1.000000   
50%    1.706996e+06  102925.000000      12.000000       3.000000   
75%    2.566178e+06  154743.000000      26.000000       5.000000   
max    3.421036e+06  206209.000000      99.000000       6.000000   

       order_hour_of_day  days_since_prior_order     product_id  \
count      202238.000000           190556.000000  202238.000000   
mean           13.181835               10.538445   18768.761009   
std             4.193794                8.380235   13345.840890   
min             0.000000                0.000000    1940.000000   
25%            10.000000                5.000000    5785.000000   
50%            13.000000                7.000000   19820.000000   
75%            16.000000               14.000000   32478.000000   
max            23.000000               30.000000   49319.000000   

       add_to_cart_order      reordered     Unnamed: 0       aisle_id  \
count      202238.000000  202238.000000  202238.000000  202238.000000   
mean            5.423051       0.782692   18769.890520      83.716626   
std             5.397505       0.412415   13347.113649       6.511429   
min             1.000000       0.000000    1940.000000      21.000000   
25%             2.000000       1.000000    5785.000000      84.000000   
50%             4.000000       1.000000   19821.000000      84.000000   
75%             7.000000       1.000000   32480.000000      84.000000   
max           137.000000       1.000000   49323.000000      91.000000   

       department_id         prices      max_order  
count       202238.0  202238.000000  202238.000000  
mean            16.0       9.086900      35.834769  
std              0.0       3.064869      26.069071  
min             16.0       1.200000       3.000000  
25%             16.0       7.900000      14.000000  
50%             16.0       9.200000      30.000000  
75%             16.0      11.400000      51.000000  
max             16.0      14.700000      99.000000

# 2% Reduced Fat Milk 
Mean = 9.01
Min = 1.2
Max = 14.7
product id = 33664 will change to 9.99 from 99,999


In [37]:
# Create a filter for products that contain 'Cottage Cheese', '2' and 'Milkfat' in the name, in dairy_without_weird
cottage_cheese = dairy_without_weird.query('product_name.str.contains("Cottage Cheese") & product_name.str.contains("2") & product_name.str.contains("Milkfat")')

In [38]:
print('Sample output from cottage_cheese:')
cottage_cheese.sample(30)

Sample output from cottage_cheese:


order_id  user_id eval_set  order_number  order_dow  \
30576373   2090443   184375    prior            19          3   
16192409   2858803   170297    prior            13          4   
29899042    148184    97092    prior            53          5   
16191122   3130901    66562    prior            34          3   
16191536   2702575    99738    prior             2          1   
16190489    484210     7669    prior            16          6   
16192351    396347   164884    prior             4          5   
16191531   1625414    98707    prior             1          6   
29899204   2600757   149006    prior             2          1   
29899328    403355   188721    prior             3          2   
16191866   3363110   118927    prior             1          5   
16190469    625431     4337    prior             3          3   
16192078   1377722   137255    prior            72          3   
16192635    903490   187697    prior            30          6   
16190977   1429635    47549    prior             7          5   
16192141   2104790   141300    prior             1          5   
16190574   1957790    11934    prior             9          6   
16192170    379629   143827    prior            12          1   
29899087   2375086   109428    prior             1          6   
16191225   1577762    76520    prior             2          5   
30576202   1059677    99045    prior            36          6   
29899075     55036   107636    prior            28          5   
30576312   3406233   149033    prior            45          6   
16191612   1319259   104403    prior            14          4   
16192315   1719559   160214    prior             2          1   
16192542   2160468   179471    prior             5          2   
16191201   1914039    73741    prior            17          2   
16191810   2223942   114280    prior             1          5   
16191655    250917   108816    prior            28          1   
29899114   1568819   116984    prior            30          6   

          order_hour_of_day  days_since_prior_order  product_id  \
30576373                  0                    10.0       11203   
16192409                 15                    14.0       27790   
29899042                 10                     7.0        6852   
16191122                  8                    10.0       27790   
16191536                 13                    12.0       27790   
16190489                 17                     5.0       27790   
16192351                 10                    30.0       27790   
16191531                  8                     NaN       27790   
29899204                 15                    30.0        6852   
29899328                  7                     7.0        6852   
16191866                 21                     NaN       27790   
16190469                 20                    30.0       27790   
16192078                  9                     2.0       27790   
16192635                  8                    30.0       27790   
16190977                  9                    10.0       27790   
16192141                  4                     NaN       27790   
16190574                 23                    15.0       27790   
16192170                 12                    19.0       27790   
29899087                 15                     NaN        6852   
16191225                 14                    16.0       27790   
30576202                 20                     8.0       11203   
29899075                 11                     5.0        6852   
30576312                 10                     4.0       11203   
16191612                 13                    29.0       27790   
16192315                 18                    30.0       27790   
16192542                 12                    30.0       27790   
16191201                  9                     4.0       27790   
16191810                 15                     NaN       27790   
16191655                 15                    13.0

In [39]:
# Quick statistics in cottage_cheese
print('Quick statistics in cottage_cheese:')
cottage_cheese.describe()

Quick statistics in cottage_cheese:


order_id        user_id  order_number    order_dow  \
count  3.487000e+03    3487.000000   3487.000000  3487.000000   
mean   1.684106e+06  103778.688271     15.484084     2.849728   
std    9.785618e+05   58583.418992     16.808776     2.031540   
min    2.301000e+03      52.000000      1.000000     0.000000   
25%    8.438150e+05   55839.000000      4.000000     1.000000   
50%    1.673171e+06  107636.000000     10.000000     3.000000   
75%    2.535750e+06  151541.500000     20.000000     5.000000   
max    3.419042e+06  206072.000000     98.000000     6.000000   

       order_hour_of_day  days_since_prior_order    product_id  \
count        3487.000000             3218.000000   3487.000000   
mean           13.257241               12.898073  22348.791225   
std             4.297401                9.231128   8826.332190   
min             0.000000                0.000000    195.000000   
25%            10.000000                6.000000  11203.000000   
50%            13.000000               10.000000  27790.000000   
75%            16.000000               19.000000  27790.000000   
max            23.000000               30.000000  31030.000000   

       add_to_cart_order    reordered    Unnamed: 0     aisle_id  \
count        3487.000000  3487.000000   3487.000000  3487.000000   
mean            5.840551     0.697448  22350.199312   107.650703   
std             5.955836     0.459429   8827.244775     5.502322   
min             1.000000     0.000000    194.000000    21.000000   
25%             2.000000     0.000000  11203.000000   108.000000   
50%             4.000000     1.000000  27792.000000   108.000000   
75%             8.000000     1.000000  27792.000000   108.000000   
max            57.000000     1.000000  31032.000000   108.000000   

       department_id       prices    max_order  
count         3487.0  3487.000000  3487.000000  
mean            16.0     7.228219    30.531116  
std              0.0     1.757900    25.413902  
min             16.0     2.400000     3.000000  
25%             16.0     7.700000    11.000000  
50%             16.0     7.700000    22.000000  
75%             16.0     7.700000    43.000000  
max             16.0    11.800000    99.000000

# Lowfat 2% Milkfat Cottage Cheese
Mean = 7.23
Min = 2.4
Max = 11.8
product id = 21553 price will change to 9.99 from 14,900


In [40]:
# Replace the price '99999' with '9.99' and impute the value '14900' with the mean price from cottage_cheese
df_ords_merged['prices'] = df_ords_merged['prices'].replace({99999: 9.99, 14900: cottage_cheese['prices'].mean()})

In [41]:
#Check the statistics of the new 'prices' column in df_ords_merged
print('The statistics of the new prices column in df_ords_merged:')
df_ords_merged['prices'].describe()

The statistics of the new prices column in df_ords_merged:


count    3.240486e+07
mean     7.790965e+00
std      4.241491e+00
min      1.000000e+00
25%      4.200000e+00
50%      7.400000e+00
75%      1.130000e+01
max      2.500000e+01
Name: prices, dtype: float64

Determine if prices of products purchased by loyal customers are different than those purchased by reglar or new customers

In [42]:
# Basic statistics of df_ords_merged by 'loyalty_flag' with the updated 'prices' column, then calculate the mean, min and max of the 'prices' column
print('The mean, min and max of prices grouped by loyalty_flag with the updated prices column in df_ords_merged:')
df_ords_merged.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

The mean, min and max of prices grouped by loyalty_flag with the updated prices column in df_ords_merged:


prices           
                      mean  min   max
loyalty_flag                         
Loyal customer    7.773526  1.0  25.0
New customer      7.801190  1.0  25.0
Regular customer  7.798239  1.0  25.0

New customers spend the most for products

Regular customers spend second most

Loyal cutomers spend least

Max values have been lowered to 25

In [43]:
# Create a new column 'avg_price' containing the average price for each user_id, calculated using the groupby() and transform() functions
df_ords_merged['avg_price'] = df_ords_merged.groupby(['user_id'])['prices'].transform(np.mean)

In [44]:
# Check the output of 'user_id' and 'avg_price'
print('Sample user_id and avg_price output:')
df_ords_merged[['user_id', 'avg_price']].head(10)

Sample user_id and avg_price output:


user_id  avg_price
0        1   6.367797
1        1   6.367797
2        1   6.367797
3        1   6.367797
4        1   6.367797
5        1   6.367797
6        1   6.367797
7        1   6.367797
8        1   6.367797
9        1   6.367797

In [45]:
# Create the 'spending_flag' column based on each user's average price
df_ords_merged.loc[df_ords_merged['avg_price'] < 10, 'spending_flag'] = 'Low spender'

In [46]:
df_ords_merged.loc[df_ords_merged['avg_price'] >= 10, 'spending_flag'] = 'High spender'

In [47]:
# Count the values in the 'spending_flag' column
print('The value counts in the spending_flag column:')
df_ords_merged['spending_flag'].value_counts(dropna = False)

The value counts in the spending_flag column:


Low spender     32285150
High spender      119709
Name: spending_flag, dtype: int64

In [48]:
# Check the accuracy of the 'spending_flag' column with output from columns 'user_id', 'avg_price', and 'loyalty_flag' column
print('Sample output from columns user_id, avg_price and spending_flag columns:')
df_ords_merged[['user_id', 'avg_price', 'spending_flag']].sample(30)

Sample output from columns user_id, avg_price and spending_flag columns:


user_id  avg_price spending_flag
2029676     26984   6.594516   Low spender
13435744    86034   7.437002   Low spender
17774653   199061   7.539496   Low spender
16939624   102993   7.312586   Low spender
31421255    41243   6.980351   Low spender
6746564    164404   7.061268   Low spender
13544203    55104   7.470096   Low spender
31390827   167850   7.539706   Low spender
9588321    112862   7.914679   Low spender
9058470        59   7.176087   Low spender
26882678   144250   7.615723   Low spender
7214738     15746   6.535664   Low spender
9484687    198007   7.927687   Low spender
4154416    202951   8.409375   Low spender
31153992     4987   7.153684   Low spender
18167029    36961   7.996552   Low spender
3537304     97887   8.480000   Low spender
14377838   133330   7.471459   Low spender
16893169   118611   7.351849   Low spender
18301152   112958   7.407143   Low spender
14326657    65764   9.154241   Low spender
28522259    77485   7.702013   Low spender
1034168     29843   7.352677   Low spender
30440726    66967   7.941667   Low spender
19302738   105553   7.466667   Low spender
3690247     21330   7.223757   Low spender
2320665     47230   7.658763   Low spender
24470739   177595   7.821667   Low spender
24630297    57617   5.062201   Low spender
991407     139455   6.754867   Low spender

In [49]:
# Create a new column 'order_freq' containing the median in days_since_prior_order for each user_id, calculated using the groupby() and transform() functions
df_ords_merged['order_freq'] = df_ords_merged.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [50]:
# Check the output of 'user_id' and 'order_freq'
print('Sample user_id and order_freq output:')
df_ords_merged[['user_id', 'order_freq']].head(10)

Sample user_id and order_freq output:


user_id  order_freq
0        1        20.5
1        1        20.5
2        1        20.5
3        1        20.5
4        1        20.5
5        1        20.5
6        1        20.5
7        1        20.5
8        1        20.5
9        1        20.5

In [51]:
# Create 'order_freq_flag' column based on each user's median in days_since_prior_order
df_ords_merged.loc[df_ords_merged['order_freq'] > 20, 'order_freq_flag'] = 'Non-frequent customer'

In [52]:
df_ords_merged.loc[(df_ords_merged['order_freq'] > 10) & (df_ords_merged['order_freq'] <= 20), 'order_freq_flag'] = 'Regular customer'

In [53]:
df_ords_merged.loc[df_ords_merged['order_freq'] <= 10, 'order_freq_flag'] = 'Frequent customer'

In [54]:
# Count the values in the 'order_freq_flag' column
print('The value counts in the order_freq_flag column:')
df_ords_merged['order_freq_flag'].value_counts(dropna = False)

The value counts in the order_freq_flag column:


Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: order_freq_flag, dtype: int64

In [55]:
# Check null values
print('Records in df_ords_merged with null values in column order_freq_flag:')
df_ords_merged[df_ords_merged['order_freq_flag'].isnull()][['user_id', 'days_since_prior_order', 'order_freq', 'order_freq_flag']]

Records in df_ords_merged with null values in column order_freq_flag:


user_id  days_since_prior_order  order_freq order_freq_flag
13645692   159838                     NaN         NaN             NaN
17251990   159838                     NaN         NaN             NaN
17622767   159838                     NaN         NaN             NaN
24138593   159838                     NaN         NaN             NaN
25880002   159838                     NaN         NaN             NaN

In [62]:
# Check records of user_id 159838
print('Records of user_id 159838:')
df_ords_merged[df_ords_merged['user_id'] == 159838]

Records of user_id 159838:


order_id  user_id eval_set  order_number  orders_day_of_week  \
13645692    895835   159838    prior             1                   0   
17251990    895835   159838    prior             1                   0   
17622767    895835   159838    prior             1                   0   
24138593    895835   159838    prior             1                   0   
25880002    895835   159838    prior             1                   0   

          order_hour_of_day  days_since_prior_order  product_id  \
13645692                 17                     NaN       10749   
17251990                 17                     NaN       33401   
17622767                 17                     NaN       23695   
24138593                 17                     NaN       21334   
25880002                 17                     NaN       22198   

          add_to_cart_order  reordered  ...  prices _merge        price_range  \
13645692                  3          0  ...     5.7   both  Mid-range product   
17251990                  6          0  ...    14.8   both  Mid-range product   
17622767                  2          0  ...     4.7   both  Low-range product   
24138593                  5          0  ...    10.2   both  Mid-range product   
25880002                  1          0  ...     1.7   both  Low-range product   

          busiest_period_of_day  max_order  loyalty_flag avg_price  \
13645692         Average orders          1  New customer      7.42   
17251990         Average orders          1  New customer      7.42   
17622767         Average orders          1  New customer      7.42   
24138593         Average orders          1  New customer      7.42   
25880002         Average orders          1  New customer      7.42   

         spending_flag  order_freq order_freq_flag  
13645692   Low spender         NaN             NaN  
17251990   Low spender         NaN             NaN  
17622767   Low spender         NaN             NaN  
24138593   Low spender         NaN             NaN  
25880002   Low spender         NaN             NaN  

[5 rows x 24 columns]

In [63]:
df_ords_merged[df_ords_merged['user_id'] == 159838]['is_new_customer']

KeyError: 'is_new_customer'

In [58]:
# Check column names in df_ords_merged so far
print('Column names in df_ords_merged:')
df_ords_merged.columns

Column names in df_ords_merged:


Index(['order_id', 'user_id', 'eval_set', 'order_number', 'order_dow',
       'order_hour_of_day', 'days_since_prior_order', 'product_id',
       'add_to_cart_order', 'reordered', 'Unnamed: 0', 'product_name',
       'aisle_id', 'department_id', 'prices', '_merge', 'price_range',
       'busiest_period_of_day', 'max_order', 'loyalty_flag', 'avg_price',
       'spending_flag', 'order_freq', 'order_freq_flag'],
      dtype='object')

In [59]:
df_ords_merged.rename(columns = {'order_dow' : 'orders_day_of_week'}, inplace = True)

In [60]:
# Change is_new_customer to a more intuitive name
df_ords_merged.rename(columns={'is_new_customer': 'first_purchase'}, inplace = True)

In [61]:
# Export df_ords_merged in .pkl format
df_ords_merged.to_pickle(r'/Users/bridget/Downloads/Instacart Basket Analysis 11-2022/02 Data/Prepared Data/orders_products_merged_3.pkl')